src/Initializer/time_stepping/LtsWeights/LtsWeights.cpp

```diff
@@ -66,26 +68,249 @@ class FaceSorter {
   }
 };
 
-void LtsWeights::computeWeights(PUML::TETPUML const &mesh, double maximumAllowedTimeStep) {
-  logInfo(seissol::MPI::mpi.rank()) << "Computing LTS weights.";
+double computeLocalCostOfClustering(const std::vector<int>& clusterIds,
+                               const std::vector<int>& cellCosts,
+                               unsigned int rate,
+                               double wiggleFactor,
+                               double minimalTimestep) {
+  assert(clusterIds.size() == cellCosts.size());
+  double cost = 0.0;
+  for (auto i = 0U; i < clusterIds.size(); ++i) {
+    const auto cluster = clusterIds[i];
+    const auto cellCost = cellCosts[i];
+    const double updateFactor = 1.0 / (std::pow(rate, cluster) * wiggleFactor * minimalTimestep);
+    cost += updateFactor * cellCost;
+  }
+
+  return cost;
+}
+
+double computeGlobalCostOfClustering(const std::vector<int>& clusterIds,
+                                     const std::vector<int>& cellCosts,
+                                     unsigned int rate,
+                                     double wiggleFactor,
+                                     double minimalTimestep,
+                                     MPI_Comm comm,
+                                     int root) {
+  double cost =
+      computeLocalCostOfClustering(clusterIds, cellCosts, rate, wiggleFactor, minimalTimestep);
+#ifdef USE_MPI
+  if (root >= 0) {
+    MPI_Reduce(MPI_IN_PLACE, &cost, 1, MPI_DOUBLE, MPI_SUM, root, comm);
+
+  } else {
+    MPI_Allreduce(MPI_IN_PLACE, &cost, 1, MPI_DOUBLE, MPI_SUM, comm);
+  }
+#endif // USE_MPI
+
+  return cost;
+}
+
+std::vector<int> enforceMaxClusterId(const std::vector<int>& clusterIds, int maxClusterId) {
+  auto newClusterIds = clusterIds;
+  assert(maxClusterId >= 0);
+  std::for_each(newClusterIds.begin(), newClusterIds.end(), [maxClusterId](auto& clusterId) {
+    clusterId = std::min(clusterId, maxClusterId);
+  });
+
+  return newClusterIds;
+}
+
+// Merges clusters such that new cost is max oldCost * allowedPerformanceLossRatio
+int computeMaxClusterIdAfterAutoMerge(const std::vector<int>& clusterIds,
+                                      const std::vector<int>& cellCosts,
+                                      unsigned int rate,
+                                      double maximalAdmissibleCost,
+                                      double wiggleFactor,
+                                      double minimalTimestep) {
+  double maxClusterId = *std::max_element(clusterIds.begin(), clusterIds.end());
+#ifdef USE_MPI
+  MPI_Allreduce(MPI_IN_PLACE, &maxClusterId, 1, MPI_DOUBLE, MPI_MAX, MPI::mpi.comm());
+#endif
+
+  // Early return
+  if (rate == 1) {
+    return maxClusterId;
+  }
+
```

Comment:
Yes, but this counting down and then returning the current id + 1 took me some time to get what is done. And IMHO, either code should speak for itself, ad otherwise there should be a comment explaining what it does.